Support question level tests

reteps opened the pull request:
If a question uses custom grading, and doesn't set `correct_answers` beforehand, the element `test` functions would currently crash. This returns out of the test function early if the element can't generate a correct submission itself. This allows for a `server.py` to define a correct answer in `test` instead.

Closes #2388

A special consideration for `pl-drawing` -- if there is no correct submission defined, then it is unclear if (1) the question should be "ungraded" (and the `test` button in PrairieLearn should succeed), by generating a parse error or (2) the question should be "graded" (the `test` button in PrairieLearn should fail until they provide a valid, non-empty submission in `server.py`). I now defer to (2), as (1) should set `gradable=false`. Before this PR, the test harness for `pl-drawing` if it found an empty correct answer, marked that it would generate a format error on parse.

SethPoulsen commented:
what is this supposed to mean?

```
    # This element cannot grade itself, so exit early.
```

do you mean that they don't grade by simple string comparisons? because that is a very different thing than not being able to grade themself.

SethPoulsen commented:
That's makes more sense for some of them, not others.

for example
pl-order-bliocks fully grades itself, no server.py needed

reteps commented:
> That's makes more sense for some of them, not others.
> 
> for example pl-order-bliocks fully grades itself, no server.py needed

Ah, I see in the prepare of `pl-order-blocks` that it always sets `data["correct_answers"][answer_name]`. I'll remove this logic from `pl-order-blocks`.

reteps commented:
Removed logic in [137e833](https://github.com/PrairieLearn/PrairieLearn/pull/11907/commits/137e833a46a910404569f1472577d6c8ac7c1e11), and updated comment in 62562d144818184d7940a7795cfa6c91d6794936.

eliotwrobson commented on apps/prairielearn/elements/pl-multiple-choice/pl-multiple-choice.py:
```diff
@@ -665,7 +665,7 @@ def test(element_html: str, data: pl.ElementTestData) -> None:
     weight = pl.get_integer_attrib(element, "weight", WEIGHT_DEFAULT)
 
-    correct_key = data["correct_answers"].get(name, {"key": None}).get("key", None)
+    correct_key = data["correct_answers"][name].get("key", None)
```
Is this entry guaranteed to be there?

reteps replied:
AFAIK yes -- see the prepare function always setting `data["correct_answers"]`.

eliotwrobson commented on exampleCourse/questions/workshop/Lesson4_example2/server.py:
```diff
@@ -25,3 +25,15 @@ def grade(data):
+
+def test(data):
+    if data["test_type"] != "correct":
+        raise NotImplementedError
```
Is this function only called if the test type is correct? Does this not raise an error?

reteps replied:
This does raise an error if triggered via the UI -- the "incorrect" and "invalid" tests fail. This is an incremental improvement of the existing behavior -- which is to simply crash. I am not a fan of requiring instructors to implement all three -- and this work is to support #11837

nwalters512 replied:
Re: #11837, historically we've been deliberate about which questions we include in our automated tests. We should consider the same here, if only in the form of an exclusion list.

The example course serves an interesting role: it contains template questions (high-quality, gold-standard), examples of elements, questions from workshops, and more. To the extent that people are looking to any of these questions for best practices, I think we should either a) implement real tests, or b) not implement tests at all and exclude these from the automated tests you're working on.

reteps replied:
In [ded45dd](https://github.com/PrairieLearn/PrairieLearn/pull/11907/commits/ded45dd647d18e45cfc9514fd301303c65e58f87) I updated the elements so they still can generate invalid inputs for some scenarios. Then, I'll update these to support generating both "correct" and "incorrect" inputs.

reteps replied:
Updated more in [08ec836](https://github.com/PrairieLearn/PrairieLearn/pull/11907/commits/08ec83674631cfbf692eb050cfe13573f9c96c46).

nwalters512 commented on apps/prairielearn/elements/pl-drawing/pl-drawing.py:
```diff
@@ -159,7 +159,8 @@ def prepare(element_html: str, data: pl.QuestionData) -> None:
                 obj["graded"] = False
             data["correct_answers"][name] = union_drawing_items(init, ans)
-        else:
+        # We only want to define a correct answer in certain cases
```
Nit, spelling out what these "certain cases" are? Tweak as needed.

```suggestion
        # Only save a correct answer if it contains any elements.
```

nwalters512 commented on apps/prairielearn/elements/pl-drawing/pl-drawing.py:
```diff
@@ -519,6 +520,12 @@ def test(element_html: str, data: pl.ElementTestData) -> None:
     )
+
+    if name not in data["correct_answers"]:
+        # This element cannot test itself, so
+        # defer the generation of test inputs to server.py
```
Nit

```suggestion
        # This element cannot test itself. Defer the generation of test inputs to server.py
```

nwalters512 commented on apps/prairielearn/elements/pl-drawing/pl-drawing.py:
Are there any questions that actually perform grading of `pl-drawing` in `server.py`? That seems really complex for any instructors to do.

reteps replied:
Yes.
![image](https://github.com/user-attachments/assets/a596f900-724f-48a8-99ed-349db960b7ae)

nwalters512 commented on exampleCourse/elements/clickable-image/clickable-image.py:
```diff
@@ -84,3 +89,12 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, "answers-name")
+    if data["test_type"] != "correct":
+        raise NotImplementedError
```
It seems like it wouldn't be too hard to actually define incorrect/invalid submissions here?

reteps replied:
Changed in [08ec836](https://github.com/PrairieLearn/PrairieLearn/pull/11907/commits/08ec83674631cfbf692eb050cfe13573f9c96c46).

reteps commented:
I had to squash it down to one commit so I could easily run it back through the test harness and ensure no regressions. It all seems to work.

reteps commented on apps/prairielearn/elements/pl-number-input/pl-number-input.py:
```diff
@@ -505,22 +513,44 @@ def test(element_html: str, data: pl.ElementTestData) -> None:
-            eps += random.uniform(1, 10)
-            answer = correct_answer + eps * random.choice([-1, 1])
+            # eps += random.uniform(1, 10)
+            answer = correct_answer + eps * random.choice([-1.1, 1.1])
```
This `eps +=` may trigger some cases where a student answer is extremely out of bounds, and the instructor wants to provide more detailed feedback.

eliotwrobson reviewed (approved):
Tests look reasonable now
